votor/src/consensus_pool/certificate_builder.rs

```diff
@@ -16,87 +16,288 @@ use {
+                    }
+                    let vote_type = VoteType::get_type(&msg.vote);
+                    if vote_type == vote_types[0] {
+                        bitmap0.set(rank, true);
+                    } else {
+                        assert_eq!(vote_type, vote_types[1]);
+                        bitmap1.get_or_insert(default_bitvec()).set(rank, true);
+                    }
+                }
+                Ok(signature.aggregate_with(msgs.iter().map(|m| &m.signature))?)
             }
-        }
 
-        Ok(self
-            .signature
-            .aggregate_with(messages.iter().map(|m| &m.signature))?)
+            Self::SingleVote { signature, bitmap } => {
+                for msg in msgs {
+                    let rank = msg.rank as usize;
+                    if MAXIMUM_VALIDATORS <= rank {
+                        return Err(AggregateError::ValidatorDoesNotExist(msg.rank));
+                    }
+                    let vote_type = VoteType::get_type(&msg.vote);
+                    assert_eq!(vote_type, vote_types[0]);
+                    bitmap.set(rank, true);
+                }
+                Ok(signature.aggregate_with(msgs.iter().map(|m| &m.signature))?)
+            }
+        }
     }
 
-    pub fn build(mut self) -> Result<CertificateMessage, CertificateError> {
-        let bitmap = match self.bitmap_1 {
-            None => {
-                let new_len = self.bitmap_0.last_one().map_or(0, |i| i.saturating_add(1));
-                self.bitmap_0.resize(new_len, false);
-                encode_base2(&self.bitmap_0).map_err(CertificateError::EncodeError)?
+    /// Builds a [`CertificateMessage`] from the builder.
+    fn build(self, certificate: Certificate) -> Result<CertificateMessage, BuildError> {
+        match self {
+            Self::SingleVote { signature, bitmap } => {
+                build_cert_from_bitmap(certificate, signature, bitmap).map_err(BuildError::Encode)
             }
-            Some(mut bitmap_1) => {
-                let last_one_0 = self.bitmap_0.last_one().map_or(0, |i| i.saturating_add(1));
-                let last_one_1 = bitmap_1.last_one().map_or(0, |i| i.saturating_add(1));
-                let new_length = last_one_0.max(last_one_1);
-                self.bitmap_0.resize(new_length, false);
-                bitmap_1.resize(new_length, false);
-                encode_base3(&self.bitmap_0, &bitmap_1).map_err(CertificateError::EncodeError)?
-            }
-        };
-        Ok(CertificateMessage {
-            certificate: self.certificate,
-            signature: self.signature.into(),
-            bitmap,
-        })
+            Self::Skip {
+                mut signature0,
+                bitmap0,
+                sig_and_bitmap1,
+            } => match sig_and_bitmap1 {
+                None => build_cert_from_bitmap(certificate, signature0, bitmap0)
+                    .map_err(BuildError::Encode),
```

Comment:
why do `build_cert_from_bitmap()` and `build_cert_from_bitmaps()` return different error types? they both seem to only ever return the same encoding error